RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -995,19 +995,67 @@ reco::Track TrackExtenderWithMTDT<TrackCollection>::buildTrack(const reco::Track
+          ifirst = ihitcount;
+        }
+        if (MTDDetId(hit.recHit()->geographicalId()).mtdSubDetector() == 2) {
+          ietlcount++;
+        }
       }
     }
 
+    auto ihit1 = trajWithMtd.measurements().cbegin() + ifirst - 1;
+    if (ihitcount == 1) {
+      const MTDTrackingRecHit* mtdhit = static_cast<const MTDTrackingRecHit*>((*ihit1).recHit()->hit());
+      thit = mtdhit->time();
+      thiterror = mtdhit->timeError();
+      validmtd = true;
+    } else if (ihitcount == 2 && ietlcount == 2) {
+      const MTDTrackingRecHit* mtdhit1 = static_cast<const MTDTrackingRecHit*>((*ihit1).recHit()->hit());
+      const MTDTrackingRecHit* mtdhit2 = static_cast<const MTDTrackingRecHit*>((*(ihit1 + 1)).recHit()->hit());
+      const auto& propresult =
+          thePropagator->propagateWithPath(ihit1->updatedState(), (ihit1 + 1)->updatedState().surface());
+      double etlpathlength = std::abs(propresult.second);
+      //
+      // The information of the two ETL hits is combined and attributed to the innermost hit
+      //
+      if (etlpathlength == 0.) {
+        validpropagation = false;
+      } else {
+        pathlength -= etlpathlength;
+        TrackTofPidInfo tofInfo =
+            computeTrackTofPidInfo(p.mag2(), etlpathlength, mtdhit1->time(), mtdhit1->timeError(), 0., 0., true);
+        //
+        // Protect against incompatible times
+        //
+        if ((tofInfo.dt - mtdhit2->time()) * (tofInfo.dt - mtdhit2->time()) <
+            ((tofInfo.dterror * tofInfo.dterror) + (mtdhit2->timeError() * mtdhit2->timeError())) * etlTimeChi2Cut_) {
+          //
+          // Subtract the ETL time of flight from the outermost measurement, and combine it in a weighted average with the innermost
+          // the mass ambiguity related uncertainty on the time of flight is added as an additional uncertainty
+          //
+          thiterror = 1. / (tofInfo.dterror * tofInfo.dterror) + 1. / (mtdhit2->timeError() * mtdhit2->timeError());
+          thit = (tofInfo.dt / (tofInfo.dterror * tofInfo.dterror) +
+                  mtdhit2->time() / (mtdhit2->timeError() * mtdhit2->timeError())) /
+                 thiterror;
+          thiterror = 1. / std::sqrt(thiterror);
+          LogDebug("TrackExtenderWithMTD") << "p trk = " << p.mag() << " ETL hits times/errors: " << mtdhit1->time()
+                                           << " +/- " << mtdhit1->timeError() << " , " << mtdhit2->time() << " +/- "
+                                           << mtdhit2->timeError() << " extrapolated time1: " << tofInfo.dt << " +/- "
+                                           << tofInfo.dterror << " average = " << thit << " +/- " << thiterror;
+          validmtd = true;
+        }
+      }
+    } else {
+      edm::LogInfo("TrackExtenderWithMTD")
+          << "MTD hits #" << ihitcount << "ETL hits #" << ietlcount << " anomalous pattern, skipping...";
+    }
+
     if (validmtd && validpropagation) {
```

Comment:
I am not sure I looked at all possible implications... but are you sure that it is worth computing all what is betweem L998 and L1058 when `validpropagation` is already `false` at L993?

Reply:
good remark, this was present even in the older version of the code, but indeed the amount of extra operations was smaller at that moment.